### ISSUE TITLE
Does this also work in multiplayer?

### ISSUE THREAD

**Issue by notcooler** [jedjoud10/VoxelTerrainGenerator#1]
I dont see a reason why you have added networking to it as well?

**Comment by jedjoud10** [comment]
I am currently working on multiplayer support as a matter if fact so yes. I
needed the multiplayer package since I needed the INetworkSerializable
interface from there since for some reason unity only has their
serialization system in the multiplayer package.

**Comment by notcooler** [comment]
Btw, how does the seaming between chunks work can you show me which script handles that?

**Comment by jedjoud10** [comment]
Since I use surface nets, the mesh vertices generated by the mesher might clip outside the boundary of a single chunk, so I just clip the vertices _near_ chunk boundaries to the actual chunk boundaries. The script that handles this is the VertexJob file. I also extend the mesh downwards (towards the terrain) using skirts, so that you can't see the sky sphere or sky box within small gaps between chunks/

https://github.com/jedjoud10/VoxelTerrainGenerator/blob/9813539090854db2c734bd5df1800e079d7e21d0/Runtime/Mesher/VertexJob.cs#L155-L179
https://github.com/jedjoud10/VoxelTerrainGenerator/blob/9813539090854db2c734bd5df1800e079d7e21d0/Runtime/Mesher/QuadJob.cs#L155-L158
https://github.com/jedjoud10/VoxelTerrainGenerator/blob/9813539090854db2c734bd5df1800e079d7e21d0/Runtime/Mesher/QuadJob.cs#L139-L142


Normally for surface nets meshes you would use something like a full octree to be able to generate "seam" meshes that make the transition between chunks more smooth, but for now this method not only is very easy to implement but also doesn't introduce inter-chunk dependencies. I will improve it in the long run though because the seams are pretty noticeable in some cases.

On the topic of skirt generation, I should note that I had an idea for generating better skirts by first creating an SDF of the voxel terrain and then only generating the skirt vertices when getting a certain distance away from the isosurface. At the moment it is checking the density value instead, which fluctuates a lot and isn't really representative of the closest distance to the surface since the domain can be warped in many ways in the compute shader.

**Comment by jedjoud10** [comment]
As for multiplayer, I already have a crude implementation working (being able to load the same world and rendering props on all players) but nothing major so far. Next steps would be to send world information (seed value and parameters) to all clients and handle synchronized voxel/dynamic edits

**Comment by notcooler** [comment]
So you just offset the vertices at the edge a bit further without knowing the data of the neighbour chunks right?

**Comment by jedjoud10** [comment]
Yes. It works most of the times but in some cases it creates "planar" surfaces with normals that differ too much from the actual surface, causing them to look very weird.

**Comment by jedjoud10** [comment]
@notcooler since I am coming back to this project I just want to let you know that I have managed to implement proper skirts with surface nets (like the ones that one would implement with marching cubes). This new system is heaps better than my old system, and the skirts are literally just not visible (compared to the old shitty ones)

![Image](https://github.com/user-attachments/assets/09691571-1563-4846-850b-f96812623b44)

![Image](https://github.com/user-attachments/assets/7547e604-caeb-434f-a489-d6f99737ebe2)


terrain with skirts that are visible (visible due to DDX DDY normals in shader)
![Image](https://github.com/user-attachments/assets/f353c9c3-6976-405a-9f8a-a782580adf21)


same exact terrain but a tweak to the shader so that skirts use _voxel_ normals instead of DDY DDX flat normals
![Image](https://github.com/user-attachments/assets/84e44570-41a7-47ee-b13a-83d77a90d8c0)

Completely different than my old method, but this one is more complex (I have to run 2D surface nets on the 6 chunk faces, and 1D surface nets on each of the 12 chunk _edges_ (for the corners I just always put a vertex there))
